### DISCUSSION TITLE
Create alias of iterator-variable in x-for.

### DISCUSSION THREAD

**Discussion by eensander** [alpinejs/alpine#1188] [1. Help]
Hi,

I am running into some problems on a very particular situation.
I tried implementing a recursive function similar as demonstrated here: https://codepen.io/ScottWindon/pen/RwabppB .

The key part of this solution is in the function renderLevel, iterating through a property of an object and assigning the iterator to the same name as object, if that makes sense:

```js
<template x-for='(level,i) in level.children'>
    <li x-html="renderLevel(level,i)"></li>
</template>
```

Now, this solution works for my problem too. Though in my situation I need to access some properties of the parent node in the looping sequence. `parent_node` should point to the parent `node` at the level of `node.children` somehow. 

```html
<template x-for="node in node.children">         <-- I want to access this node (from node.children)
    <div>
        <div x-html="render_sub(node)"></div>
        <span x-text="parent_node.name"></span>      <-- from here
    </div>
</template>
```

I think, if thinking in traditional (js) terms, I need to create a alias or copy of the `node` at the level where `node.dynamic.node_stack` occurs, like:

```js
node = {name: 'A', 'children': [{name: 'B'}]};
function recursive(node) {
    let parent_node = node;             // creating copy assignment
    for node in node.children {
        console.log(parent_node.name);  // now I can access both the parent, and child node.
    }
}
```

I don't know if manually assigning variables at certain levels in the DOM-structure is possible in (alpine)js. 

I also understand that the solution from codepen isn't beautiful and probably not how alpinejs intended it, but for now it seems to fit my solution best. 

Any (other) possible solution to this case would be **greatly** appreciated.

**Comment by KevinBatdorf** [comment]
Do you have a demo of your code? Note that `<template>` tags can only have one child element, so the `<span>` in your example won't be rendered and you will get a warning in the console.

**Reply by eensander** [reply]
Thanks for your reply.

Well, the application this code applies to is relatively complex so I simplified it a little bit. In this simplification I indeed forgot to add an enclosing div to those two elements. I updated my question respectively. 

The code from which the given example is derrived comes down to:

```html
                            <template x-for="(node, node_index, parent_stack) in node.dynamic.node_stack">
                            ...
                                        <!-- the recursive rendering of the child node(s): -->
                                            <template x-for="(node, node_index) in node.dynamic.node_stack">
                                                <div x-html="render_subgraph_node(node)"></div>
                                            </template>
                            ...
                                        <!-- the need to access the parent node: -->
                                        <div
                                                @click="subgraph_click_node_option(parent_node, node, edge_out)"
                                                class="mx-1 px-2 py-1 inline-block "
                                                :class="classes_edge(node, edge_out)">

                                                <span class="select-none" x-text="edge_out.content ?? 'Continue'"></span>

                                        </div>
                            ...
                            </template>
```

AlpineJS's relevant data used in the example above (node_stack) looks like (also simplified):

```js
node_stack: [
    {
        id: 1,
        content: 'start',
        dynamic: {}
    },
    {
        id: 2,
        content: 'question',
        dynamic: {
            node_stack: [
                {
                    id: 3,
                    content: 'yes',
                    dynamic: {}
                },
                {
                    id: 4,
                    content: 'no',
                    dynamic: {}
                },
            ]
        }
    },
    {
        id: 5,
        content: 'end',
        dynamic: {}
    },
]
```

I hope this clears it up a little.

**Reply by KevinBatdorf** [reply]
Still kind of hard to tell what you need with only seeing pieces of it, I think Are you trying to get a node that's a direct parent or something else? Have you tried just using vanilla JavaScript? `.closest()` might be what you're looking for? 

If you're doing recursion you may need to keep track of where you are in the stack too. It just depends.

Anyway you can put a small demo on codepen showing it?

**Reply by eensander** [reply]
> Still kind of hard to tell what you need with only seeing pieces of it, I think Are you trying to get a node that's a direct parent or something else? Have you tried just using vanilla JavaScript? `.closest()` might be what you're looking for?
> 
> If you're doing recursion you may need to keep track of where you are in the stack too. It just depends.
> 
> Anyway you can put a small demo on codepen showing it?

Thanks for your replies, but because of this case being so specific I converted this component to Vue which helped me reach my goal easier, due to it's inherent capabilities of making components recursive. 

Once again, thanks for your time.
